app/src/main/java/com/example/drawingactivity/GradientView.kt

```diff
@@ -0,0 +1,163 @@
+                    onSatValChanged?.invoke(dx / width, dy / height)
+                    invalidate()
+                }
+                MotionEvent.ACTION_MOVE, MotionEvent.ACTION_CANCEL -> {
+                    dx = mEvent.x.coerceIn(0f, width * 1f)
+                    dy = mEvent.y.coerceIn(0f, height * 1f)
+                    if (dy > height / 2) {
+                        selectPaint.color = Color.WHITE
+                    } else selectPaint.color = BLACK
+                    onSatValChanged?.invoke(dx / width, dy / height)
+                    invalidate()
+                }
+                MotionEvent.ACTION_UP -> {
+                    onGradientClicked?.invoke(false)
+                }
+            }
+            true
+        }
+    }
+
+    @OptIn(DelicateCoroutinesApi::class)
+    private fun generateAllBitmaps() {
+        if (!arrayIsFull) {
+            GlobalScope.launch {
+                withContext(Dispatchers.Default) {
+                    for (i in 0..90) {
+                        bitmaps!![i] = makeBitmap(i.toFloat())
+                    }
+                }
+            }
+            GlobalScope.launch {
+                withContext(Dispatchers.Default) {
+                    for (i in 91..180) {
+                        bitmaps!![i] = makeBitmap(i.toFloat())
+                    }
+                }
+            }
+            GlobalScope.launch {
+                withContext(Dispatchers.Default) {
+                    for (i in 181..270) {
+                        bitmaps!![i] = makeBitmap(i.toFloat())
+                    }
+                }
+            }
+            GlobalScope.launch {
+                withContext(Dispatchers.Default) {
+                    for (i in 271..360) {
+                        bitmaps!![i] = makeBitmap(i.toFloat())
+                    }
+                    arrayIsFull = true
+                }
+            }
+        }
+    }
+
+    private fun makeBitmap(hue: Float): Bitmap? {
+        val step = 1f / imageSize
+        var v = 1f
+        var s = 0f
+        val hsv = floatArrayOf(hue, 0f, 0f)
+        var index = 0
+        val pixels = IntArray(imageSize * imageSize)
+
+        repeat(imageSize) {
+            repeat(imageSize) {
+                hsv[1] = s
+                hsv[2] = v
+                pixels[index] = HSVToColor(hsv)
+                s += step
+                index += 1
+            }
+            s = 0f
+            v -= step
+        }
+        return Bitmap.createBitmap(pixels, imageSize, imageSize, Bitmap.Config.ARGB_8888)
+    }
+
+    private fun getBitmap(index: Int): Bitmap? {
+        return if (bitmaps?.get(index) == null) {
+            bitmaps?.set(index, makeBitmap(index.toFloat()))
+            bitmaps?.get(index)
+        } else bitmaps!![index]
+    }
+
+    fun setHue(huePickerValue: Int) {
+        hue = huePickerValue
+        invalidate()
+    }
+
+    fun setSaturationAndValue(x: Float, y: Float) {
```

Comment:
Parameters should be named what they are

Reply:
fixed